hail/python/hailtop/aiotools/fs.py

```diff
@@ -130,13 +130,30 @@ async def remove(self, url: str) -> None:
         pass
 
     @abc.abstractmethod
-    async def rmtree(self, sema: asyncio.Semaphore, url: str) -> None:
+    async def rmtree(self, url: str, sema: Optional[asyncio.Semaphore] = None) -> None:
         pass
 
     async def touch(self, url: str) -> None:
         async with await self.create(url):
             pass
 
+    async def read(self, url: str) -> bytes:
+        async with await self.open(url) as f:
+            return await f.read()
+
+    async def read_from(self, url: str, start: int) -> bytes:
+        async with await self.open_from(url, start) as f:
+            return await f.read()
+
+    async def read_range(self, url: str, start: int, end: int) -> bytes:
+        n = (end - start) + 1
+        async with await self.open_from(url, start) as f:
+            return await f.read(n)
+
+    async def write(self, url: str, data: bytes, *, retry_writes: bool = True) -> None:
+        async with await self.create(url, retry_writes=retry_writes) as f:
+            await f.write(data)
```

Comment:
So if you have all the data present, you shouldn't take a retry_writes argument.  You should do the rewrite with `retry_writes=False` and catch and retry yourself like it is done in `_copy_file_multi_part_main`:

```
    async def _copy_file(self, srcfile: str, destfile: str) -> None:
        ...
        dest_cm = await self.router_fs.create(destfile, retry_writes=False)
        async with dest_cm as destf:
            ...

    async def _copy_file_multi_part_main(...):
        ...
        await retry_transient_errors(self._copy_file, srcfile, destfile)
```

Reply:
So if create isn't retrying writes, you need to retry writes.  I was thinking of something like this:

```
    async def write(self, url: str, data: bytes) -> None:
        async def _write() -> None:
            async with await self.create(url, retry_writes=False) as f:
                await f.write(data)

        await retry_transient_errors(_write)
```

Let me clarify the meaning of retry writes.  By default, AsyncFS.create is retry safe.  To be retry safe with writes, the AsyncFS code has to buffer any data sent to GCS (S3, etc.) until it is acknowledged and safely stored.  That's not free.  If you pass `rewrite_retries=False`, AsyncFS.create doesn't retry writes, in which case it can fail.  `retry_writes=False` assumes that you're retrying failures at a higher level.  That's possible when you can restart the write from scratch.  That's the case here, because `data` is bytes, so if the write failed, we already have the data buffered and just write again.

Reply:
That makes more sense. Thanks!